### PR TITLE
Add sampleoutput.h parser

### PR DESCRIPTION
**Pull request by seldauyanik-maxim**
Sample notebook for parsing izer generated ampleoutput.h file.

Sample numpy input is fed to the PyTorch model that is read from training repository and values are compared with the memeory cotents logged in the sampleoutput.h file.

Sample notebook is designed for verificaiton of SVHN dataset and tinier-ssd model.

### CODE REVIEW

**Review by MaximGorkem — DISMISSED** [review]
Looks nice

File: utils/parse_sampleoutput.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    return out_dict
+
+
+def process_four_channel_data_from_word(word_value):
+    """
+    For 8 bit outputs, izer combines 4 of them and places all into a single memory word.
+    This function seperates out the 4 channels from a given word value.
+    """
+
+    ch0 = int(word_value, 16) & int('0x000000ff', 0)
```
Comment (rotx-maxim) [inline utils/parse_sampleoutput.py:81]: int('0x000000ff', 0)

Use 0x000000ff instead?

File: utils/parse_sampleoutput.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                addr = None
+                d_len = None
+                mask = None
+
+    return out_dict
+
+
+def process_four_channel_data_from_word(word_value):
+    """
+    For 8 bit outputs, izer combines 4 of them and places all into a single memory word.
```
Comment (rotx-maxim) [inline utils/parse_sampleoutput.py:77]: Note that word_value must be a hexadecimal string

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3) Compare PyTorch Output with the sampleoutput.h Contents\n",
+    "\n",
+    "Highly dependent on the yaml file: networks/svhn-tinierssd.yaml\n",
+    "\n",
+    "Important parts: \n",
+    "* wide, non-wide outputs\n",
+    "* Concatenated addresses in sampleoutput.h (Check yam for output addresses of output layers)\n"
```
Comment (rotx-maxim) [inline notebooks/Parse_sampleout_and_compare_SVHN.ipynb:339]: yam -> yaml

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+    "# Layer 17: cl_conv9 (81, 44),   0x3510 - 0x3654\n",
+    "# Layer 18: cl_conv10 (16, 44),   0x3654 - 0x3694\n",
+    "# Layer 19: cl_conv12_2 (4, 44), 0x3694 - 0x36A4\n",
+    "# are all concetenated using most significant 44 processors\n",
+    "address_offset = '0x3000'\n",
+    "wide = False\n",
+    "output_shape = (425, 44)\n",
+    "block_1 = (address_offset, wide, output_shape)\n",
+    "\n",
+    "# IMPORTANT: For MAX78000, not all output offfsets will reside in different xxxx3000 addresses\n",
```
Comment (rotx-maxim) [inline notebooks/Parse_sampleout_and_compare_SVHN.ipynb:384]: offsets

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+    "block = blocks[block_id]\n",
+    "offset = block[0]\n",
+    "wide = block[1]\n",
+    "output_shape = block[2]\n",
+    "read_block_1 = np.zeros(output_shape)\n",
+    "\n",
+    "read_blocks.append(read_block_0)\n",
+    "read_blocks.append(read_block_1)\n",
+    "\n",
+    "# PArse sample_output_h_dict items:\n",
```
Comment (rotx-maxim) [inline notebooks/Parse_sampleout_and_compare_SVHN.ipynb:421]: Parse

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+    "\n",
+    "    block = blocks[block_id]\n",
+    "    offset = block[0]\n",
+    "    wide = block[1]\n",
+    "    output_shape = block[2]\n",
+    "\n",
+    "    num_of_channels =  output_shape[1]\n",
+    "    num_of_groups = math.ceil(num_of_channels / 4)\n",
+    "\n",
+    "    # This if clause if true for all entires of this sample yaml\n",
```
Comment (rotx-maxim) [inline notebooks/Parse_sampleout_and_compare_SVHN.ipynb:437]: is true?

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+    "    # TODO: you may search key ends with the offset read from yamfile instead\n",
+    "    if key.endswith(offset[2:]) or key.endswith(hex(int(offset[2:], 16) + int('0x8000', 16))[2:]):\n",
```
Comment (rotx-maxim) [inline notebooks/Parse_sampleout_and_compare_SVHN.ipynb:439]: int('0x8000', 16) -> 0x8000

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
Just some small changes

File: utils/parse_sampleoutput.py (hunk 1 above)
Comment (seldauyanik-maxim) [reply utils/parse_sampleoutput.py:81]: Done

File: utils/parse_sampleoutput.py (hunk 2 above)
Comment (seldauyanik-maxim) [reply utils/parse_sampleoutput.py:77]: Done

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb (hunk 3 above)
Comment (seldauyanik-maxim) [reply notebooks/Parse_sampleout_and_compare_SVHN.ipynb:339]: Done

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb (hunk 4 above)
Comment (seldauyanik-maxim) [reply notebooks/Parse_sampleout_and_compare_SVHN.ipynb:384]: Done

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb (hunk 5 above)
Comment (seldauyanik-maxim) [reply notebooks/Parse_sampleout_and_compare_SVHN.ipynb:421]: Done

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb (hunk 6 above)
Comment (seldauyanik-maxim) [reply notebooks/Parse_sampleout_and_compare_SVHN.ipynb:437]: Done

File: notebooks/Parse_sampleout_and_compare_SVHN.ipynb (hunk 7 above)
Comment (seldauyanik-maxim) [reply notebooks/Parse_sampleout_and_compare_SVHN.ipynb:439]: Done

**Review by rotx-maxim — APPROVED** [review]

**Pull request merged**